pull sync: initial version

acud opened the pull request:
### Review guide

_NOTE: This PR has been splitted into 2 commits for easier review process._

There are several components to be reviewed in this PR:
- `pullsync` protocol package - exposes APIs to synchronize given bins and intervals from a peer (implemented in first commit)
- `pullstorage` package - wraps a `localstore` instance to provide functionality to create batches from requested intervals in a bin (implemented in first commit)
- `puller` package - in an analogy to `pushsync` and `pusher`, I have created the `puller` component which is decoupled from `pullsync` package, and is tasked with syncing with peers. It receives notifications from the topology whenever a peer connects or disconnects, and similarly to the `manage` loop in kademlia that prioritizes peers to connect to, it prioritizes peers to start syncing with or to recalculate subscriptions with (when depth changes) (implemented in first commit)

My suggestion would be to start with reviewing the `pullsync` and `pullstorage` packages in the first commit as they are dependent upon each other (`pullstorage` is a `pullsync` dependency), then change to the second commit which introduces further changes to these packages (although pretty minor), and the `puller` component.

The PR tries to employ the principles that have been guiding us so far with `bee` development, thus mocks are heavily used across the packages. Inadvertently, since the real-world scenarios are quite complex (and ... deadlines) resulted in some of the mocks having quite some business logic in order to reproduce the desirable behavior needed for the test. This can be refactored later on, and I am aware of some of it being very verbose and with overlapping functionalities. I am also aware that it might not be entirely trivial for those who have not worked with this package in the old codebase to understand why a certain behavior is needed from the mock in a given test case. If this happens - please comment and ask in the PR or on mattermost.

This code has been thoroughly compared with all of the nuts and bolts in the stream package in `swarm` and it seems to cover the edge cases the we caught there (minus the localstore leakage fix that was reverted into the bee codebase @ #302).

One notable difference with the `swarm` stream implementation is that now peer initiation is made purely from changes in the connectivity. In `swarm`, each protocol is invoked with a `Run` function that bootstraps the protocol, and in the stream protocol the protocol would process each peer individually through the `Run` function. This causes all sorts of complexities in the protocol that must maintain a lot of state about peers. In the implementation presented in this PR peers are notified upon through the topology (i.e. whenever a change in peers occurs `puller` is notified, but does not get notified about what exactly happened but that a change has occurred, implying that peers should be recalculated). 

The depth notification from `kademlia` has been changed with a peer collection notification, which is more generalized but suits the needs of the protocol better. A simple example might be that a peer was added, but the depth did not change, and we would like to sync with that peer for whatever reason. Obviously this is needed since the abstraction does not contain the dreadful `Run` entry point.

---------

EDIT: Found bug where a peer that disconnects would not get removed correctly from `puller`. This is now fixed

---------
- [x] import bitvector from `swarm`
- [x] add db close channel to SubscribePull to signal db closed when iterator returns
- [x] implements only the dumb protocol that handles the message exchanges
- [x] implement `pullstorage` module that wraps `storage.Storer` and handles basic operations of getting, putting, setting chunks from the localstore in return for requests coming in from the protocol
- [x] implement puller
- [x] kademlia peer change subscriptions (done)
- [x] cursor exchange
- [x] handle peer slice change
- [x] handle depth changes
- [x] kademlia peer change subscription test disconnect trigger
- [x] integration testing with real localstore on pull sync
- [ ] falling edge kademlia channel wrapper -> defer to later pr
- [x] batch timeout test (for timer added code in pullstorage)
- [x] end to end context cancellation - currently only upstream peer quits but in fact there has to be some e2e closure of the stream so that downstream quits too, need to change to fullclose there
- [x] get rid of the localstore gc hack i introduced
- [x] integration testing
- [ ] metrics - later pr
- [ ] todo - check no goroutine/mem leaks
- [x] todo - peer that disconnects does not result in sync context removal from map. this can result in a mem leak and bugs when the peer gets reconnected

pradovic commented:
Looks good to me as well so far, no special comments. I will wait for ready for review, and go through it again once is ready for review. I agree about bitvector that it could be separate PR, but also I don't think this PR is huge, as @janos said, large part is auto-generated code, so it's seems fine!

acud commented:
thanks guys. I removed the deliberate panic in an earlier commit today. i now also removed the implicit methods of Has and Set. I left the Get that invokes GetMulti and the Put, but yeah we can also get rid of those too. I could maybe take care of it when the other PRs are merged into this one? just don't wanna deal too much with rebasing one on top of the other.

pradovic commented on pkg/bitvector/bitvector.go:
```diff
@@ -0,0 +1,124 @@
+}
+
+// New creates a new bit vector with the given length
+func New(l int) (bv *BitVector, err error) {
```
🌷 
```suggestion
func New(l int) (*BitVector, error) {
```

pradovic commented on pkg/bitvector/bitvector.go:
```diff
@@ -0,0 +1,124 @@
+// NewFromBytes creates a bit vector from the passed byte slice.
+//
+// Leftmost bit in byte slice becomes leftmost bit in bit vector
+func NewFromBytes(b []byte, l int) (bv *BitVector, err error) {
```
🌷 
```suggestion
func NewFromBytes(b []byte, l int) (*BitVector, error) {
```

pradovic commented on pkg/kademlia/kademlia.go:
```diff
@@ -377,6 +399,44 @@ func (k *Kad) ClosestPeer(addr swarm.Address) (swarm.Address, error) {
+	channel := make(chan struct{}, 1)
+	var closeOnce sync.Once
+
+	k.peerSigMtx.Lock()
+	defer k.peerSigMtx.Unlock()
+
+	k.peerSig = append(k.peerSig, channel)
```
🌷 Maybe it's a little bit nicer as the function below also locks, so someone might make a mistake while changing stuff.
```suggestion
	k.peerSigMtx.Lock()
	k.peerSig = append(k.peerSig, channel)
        k.peerSigMtx.Unlock()
```

acud replied:
i don't think this is too confusing tbh and i prefer the `defer` statement

acud commented on pkg/pullsync/pullstorage/pullstorage.go:
```diff
@@ -0,0 +1,147 @@
+
+// Cursors gets the last BinID for every bin in the local storage
+func (s *ps) Cursors(ctx context.Context) (curs []uint64, err error) {
+	curs = make([]uint64, 16)
```
this needs to be extracted (16)

acud replied:
this needs to be extracted

pradovic commented on pkg/kademlia/mock/kademlia.go:
```diff
@@ -0,0 +1,175 @@
+}
+
+func (m *Mock) Close() error {
+	panic("not implemented") // TODO: Implement
```
I think that comment is not needed here as the string in panic is self-desribing

acud replied:
this is auto generated text from `:GoImpl` (`vim-go`)

pradovic commented on pkg/puller/puller.go:
```diff
@@ -0,0 +1,447 @@
+			})
+
+			if err != nil {
+				panic(err)
```
Why panic here?

acud commented on pkg/pullsync/pullstorage/pullstorage_test.go:
```diff
@@ -0,0 +1,376 @@
+		swarm.MustParseHexAddress("0005"),
+		swarm.MustParseHexAddress("0006"),
+	}
+	limit = 5
```
the limit here at the tests is injected through the function call. however, pull sync protocol enforces its own limit which is different (right now set at 50), some of the test cases around the pull sync protocol are also checking around the limit (of 50 - `pullsync.go#L37`) and so this is a bit of a debt that needs to be settled. up to you guys if this should be during this PR or not

pradovic commented on pkg/pullsync/pullsync.go:
```diff
@@ -0,0 +1,333 @@
+			bv.Set(i / swarm.HashSize)
+		}
+	}
+	fmt.Println("sending want", len(bv.Bytes()))
```
loggger or maybe debug leftover?

acud commented on pkg/puller/puller.go:
```diff
@@ -0,0 +1,449 @@
+			// wait and retry? this is a local error
+			// maybe just quit the peer entirely.
+			// not sure how to do this
+			<-time.After(30 * time.Second)
```
probably log error and return here. not sure.

acud commented on pkg/puller/puller.go:
```diff
@@ -0,0 +1,449 @@
+			return
+		}
+		if top == 0 {
+			return //TODO need to deal with this somehow. not right
```
this is not legal. should not happen. probably log and return.

zelig commented on pkg/pullsync/pullsync.go:
```diff
@@ -0,0 +1,332 @@
+}
+
+func (s *Syncer) setChunks(ctx context.Context, addrs ...swarm.Address) error {
+	return s.storage.Set(ctx, storage.ModeSetSyncPull, addrs...)
```
I dont thin we need ModeSetSyncPull any more. it was legacy

acud replied:
i dont remember this was defined anywhere. can we iterate on this?

acud commented on pkg/localstore/subscription_pull.go:
```diff
@@ -36,7 +36,7 @@ import (
 // Make sure that you check the second returned parameter from the channel to stop iteration when its value
 // is false.
-func (db *DB) SubscribePull(ctx context.Context, bin uint8, since, until uint64) (c <-chan storage.Descriptor, stop func()) {
+func (db *DB) SubscribePull(ctx context.Context, bin uint8, since, until uint64) (c <-chan storage.Descriptor, closed <-chan struct{}, stop func()) {
```
theoretically yes. in the old codebase for some reason though this wasn't so for some reason and you could get some empty batch on shutdown. the problem is also that `SubscribePull` does not provide an explicit signal about shutdown (before this change). I rather be on the safe side for now. But theoretically yes you are right and it could be that the entire section of code here which is handling shutdown could be removed :+1: Need to test this though to verify the assumption.

acud commented on pkg/puller/puller.go:
```diff
@@ -0,0 +1,501 @@
+
+		cursors: make(map[string][]uint64),
+
+		syncPeers: make([]map[string]*syncPeer, bins),
```
well not entirely. you need to maintain the contexts of which peers you sync with on which bins. I wouldn't like to overload this information in kademlia

zelig replied:
you prefer replicating the container.  ok

acud replied:
i'd like to be leave the flexibility to maintain many connections in kademlia, but not sync to all of them, so maintaining a subset of the state is mandatory in this case (or flagging them somehow in kademlia, which we agreed we shouldn't do)

acud commented on pkg/puller/puller.go:
```diff
@@ -0,0 +1,501 @@
+)
+
+var (
+	bins = uint8(16)
```
well actually not a const, since i'm altering this value in tests (to make the tests more readable), it has to be a `var` in order to do so. but it could be that it would just be nicer to inject this value through the constructor. there's a few other things that could follow

acud commented on pkg/puller/puller.go:
```diff
@@ -0,0 +1,501 @@
+				syncing := len(bp)
+				switch {
+				case po < depth:
+					// outside of depth, sync peerPO bin only
```
i remember about not syncing bin `0`. we can disable this of course.

zelig replied:
yes please apply the minimal  design we so successfully kept so far. pullsync needed now for  only neighbourhood replication and adjustment (no need for filling in storage functionality now) or trickle chunks we got pushsync protocol

acud commented on pkg/puller/puller.go:
```diff
@@ -0,0 +1,501 @@
+		quit:      make(chan struct{}),
+	}
+
+	for i := uint8(0); i < bins; i++ {
```
not sure i understand what you mean

zelig replied:
u  get clleaner code if you  hide the particulars of syncCtx and just call a function on it

acud commented on pkg/puller/puller.go:
```diff
@@ -0,0 +1,501 @@
+		// outside of depth
+		var (
+			want     = po
+			dontWant = []uint8{0} // never want bin 0
```
because we said we never want to pull sync on bin `0`

zelig replied:
this is not so,, i remember clearly  our call when we went through this. lets call

acud commented on pkg/puller/puller.go:
```diff
@@ -0,0 +1,501 @@
+			// maybe blacklist for some time
+		}
+		p.cursorsMtx.Lock()
+		p.cursors[peer.String()] = cursors
```
i dont understand. you need to keep a copy of the cursors in case you'd like to start syncing on some bins later on (if the depth moves back and forth). you don't want to ask the cursors if you already asked for them before

zelig replied:
are they stored in db and here you just  cache them after query? then ok

acud replied:
no they are only stored in memory (in the map...)
